base/base/defines.h

```diff
@@ -159,9 +159,9 @@
 #    define TSA_REQUIRES_SHARED(...)
 #    define TSA_NO_THREAD_SAFETY_ANALYSIS
 
-#    define TSA_SUPPRESS_WARNING_FOR_READ(x)
-#    define TSA_SUPPRESS_WARNING_FOR_WRITE(x)
-#    define TSA_READ_ONE_THREAD(x)
+#    define TSA_SUPPRESS_WARNING_FOR_READ(x) ([&]() -> const auto & { return (x); }())
```

Comment:
This code is not entirely obvious to me.

1. Why lambda is needed instead of just `x`?
2. Why do we need to call lambda and return its argument?
3. Why do we need to specify the return type for the lambda?

Let's add some comments.

Reply:
I see a similar code above.
But it's difficult to imagine that we've invented this code.
If it has been adopted from somewhere, let's add a link.

Reply:
In this case we can get away with `(x)`. I'll change it.

I think it's done above because the attribute is meant to the used for functions and those macros allows you to use it on specific statements inside a function. Maybe @rschu1ze can confirm.

Reply:
Yes, we don't need lambda here, but we need it above to apply `TSA_NO_THREAD_SAFETY_ANALYSIS` macro (which is actually a function attribute) to a statement instead of a function. It allows to suppress specific warnings locally like this:
https://github.com/ClickHouse/ClickHouse/blob/ab702e43fc574f817835c64cdccb63e8749f096d/src/IO/WriteBufferFromS3.cpp#L506-L509
instead of disabling TSA for whole function like this
https://github.com/ClickHouse/ClickHouse/blob/5365b105ccc835485215502783856158288c2b91/src/Access/DiskAccessStorage.h#L53

> Why do we need to call lambda and return its argument?

It's a way to apply function attribute to a statement. I'm pretty sure that compiler simply removes such lambdas without introducing any overhead in runtime. 

> Why do we need to specify the return type for the lambda?

To distinguish const/non-const (read/write) access. `TSA_SUPPRESS_WARNING_FOR_READ` returns constant reference.

> But it's difficult to imagine that we've invented this code.

Actually I have invented this :) Do you know a better way to use `__attribute__((no_thread_safety_analysis))` for specific statements?

Reply:
As @tavplubix wrote, TSA has no build-in way to suppress warnings at statement-level, this only works at function scope. The lambda emulates a function invocation as a workaround.

Chromium implements a similar trick: https://github.com/chromium/chromium/blob/98448449d77ed013773c2f887f15b6bced92f27f/base/thread_annotations.h#L234

I agree that a comment in the code would be nice.

Reply:
I'll update the comment for better clarity in this PR then. Thanks both for the information